synapse/storage/background_updates.py

```diff
@@ -581,22 +581,20 @@ def create_index_psql(conn: Connection) -> None:
         def create_index_sqlite(conn: Connection) -> None:
             # Sqlite doesn't support concurrent creation of indexes.
             #
-            # We don't use partial indices on SQLite as it wasn't introduced
-            # until 3.8, and wheezy and CentOS 7 have 3.7
-            #
             # We assume that sqlite doesn't give us invalid indices; however
             # we may still end up with the index existing but the
             # background_updates not having been recorded if synapse got shut
             # down at the wrong moment - hance we use IF NOT EXISTS. (SQLite
             # has supported CREATE TABLE|INDEX IF NOT EXISTS since 3.3.0.)
             sql = (
                 "CREATE %(unique)s INDEX IF NOT EXISTS %(name)s ON %(table)s"
-                " (%(columns)s)"
```

Comment:
This does mean that there is going to be drift depending on when your SQLite install go an index. It looks like there's a handful of places we use:

```sh
$ grep -r "where_clause=" synapse | grep -v "room.py"
synapse/storage/databases/main/events_bg_updates.py:            where_clause="contains_url = true AND outlier = false",
synapse/storage/databases/main/events_bg_updates.py:            where_clause="have_censored",
synapse/storage/databases/main/events_bg_updates.py:            where_clause="NOT have_censored",
synapse/storage/databases/main/events_bg_updates.py:            where_clause="contains_url = true AND outlier = false",
synapse/storage/databases/main/roommember.py:            where_clause="forgotten = 1",
synapse/storage/databases/main/presence.py:            where_clause="state != 'offline'",
synapse/storage/databases/main/media_repository.py:            where_clause="url_cache IS NOT NULL",
synapse/storage/databases/main/event_push_actions.py:            where_clause="highlight=1",
synapse/storage/databases/main/event_push_actions.py:            where_clause="highlight=0",
synapse/storage/databases/main/state.py:            where_clause="type='m.room.member'",
```

I think all of these are "simply" speed-ups, except for:

https://github.com/matrix-org/synapse/blob/53b77b203ac12f20a6534393464588d5d49435f5/synapse/storage/databases/main/media_repository.py#L70-L76

**Do we need to re-create these indices on SQLite?**

Reply:
I think for our own sanity we probably want to do that? Otherwise we're going to get very confused if anything breaks because of this

Reply:
I was very afraid you would say that. Do you think it is ok to add a "drop of exists" or should I do it all manually in a delta?

Reply:
Could we alternatively add an option to ignore the where clause on SQLite, effectively grandfathering the existing indices in?

Reply:
On the other hand, it's just an index so it's not 'possible' (modulo bugs in SQLite) for leaving the existing indices without the where clause to cause bugs.

However... how bad would it actually be to add a SQLite schema delta to rebuild the indices?
We can probably assume SQLite databases are small enough to do this without too much worry.

> I think all of these are "simply" speed-ups, ~~except for <...>~~

I think even the index you're showing is just a speed-up.
The only reason and index isn't 'just' a speed-up is when it's being used to enforce a UNIQUE constraint.

Reply:
I added a delta to re-build them. It is a little cheeky since we just drop the indices, but then rebuild them in the background. I *think* this is likely OK though. 🤷